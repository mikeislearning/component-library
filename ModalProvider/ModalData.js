// @format
import { navigate, routes } from 'app/navigation';
import { logEvent } from 'app/services';
import { images } from 'app/theme';

const getOnBoardingData = (modalProvider, notificationPressed, notificationPressedText) => {
  return [
    {
      title: 'Turn on notifications',
      image: images.localOnboardingOne,
      text: 'Never miss notifications about new messages, connections, and resources.',
      button: {
        onPress: () => {
          logEvent('onboarding_enable_notifications');
          return notificationPressed ? null : modalProvider.requestPermission();
        },
        text: notificationPressed ? notificationPressedText : 'Enable Notifications',
        theme: notificationPressed ? 'quaternary' : 'primary',
      },
      smallButton: {
        onPress: () => {
          logEvent('onboarding_skip');
          return modalProvider.closeModal();
        },
        text: 'Skip',
      },
    },
    {
      title: 'Meet your Ambassador',
      image: images.localOnboardingTwo,
      text:
        "Our Ambassadors are professionals providing career-related tips and advice specific to your field. You'll recognize Arrive Ambassadors by the gold circle on their profile.",
      button: {
        onPress: () => {
          logEvent('onboarding_meet_ambassador');
          return modalProvider.closeModal();
        },
        text: 'Start connecting',
      },
      smallButton: {
        onPress: () => {
          logEvent('onboarding_skip');
          return modalProvider.closeModal();
        },
        text: 'Skip',
      },
    },
    {
      title: 'Check your connections!',
      image: images.localOnboardingThree,
      text:
        "Every week, we match you with three new connections based on your profile. Start connecting with this week's connections today!",
      button: {
        onPress: () => {
          modalProvider.toggleVisibility(false);
          logEvent('onboarding_view_profile');
          navigate(routes.profile);
        },
        text: 'View my profile',
      },
      smallButton: {
        onPress: () => {
          logEvent('onboarding_skip');
          return modalProvider.toggleVisibility(false);
        },
        text: 'Skip',
      },
    },
  ];
};

export { getOnBoardingData };
