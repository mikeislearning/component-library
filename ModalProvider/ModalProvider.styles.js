/* eslint-disable */

import { StyleSheet } from 'react-native';

import { colors } from 'app/theme';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  contentContainer: {
    alignItems: 'center',
    backgroundColor: colors.white,
    borderRadius: 8,
    elevation: 3,
    flex: 1,
    justifyContent: 'center',
    marginHorizontal: 10,
    marginVertical: 100,
    paddingHorizontal: 20,
    paddingTop: 40,
    shadowColor: colors.black,
    shadowOffset: { width: 1, height: 1 },
    shadowOpacity: 0.2,
    shadowRadius: 2,
  },
  scrollView: {
    backgroundColor: colors.semiTransparent,
    flex: 1,
  },
});

export default styles;
