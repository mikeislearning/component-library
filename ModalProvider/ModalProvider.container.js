import { connect } from 'react-redux';
import { postFeedback } from 'app/actions';

import ModalProvider from './ModalProvider';

const mapStateToProps = (state) => ({
  error: state.feedback.error,
  isLoading: state.feedback.loading,
  isSubmitted: state.feedback.data,
});
export default connect(
  mapStateToProps,
  { postFeedback }
)(ModalProvider);
