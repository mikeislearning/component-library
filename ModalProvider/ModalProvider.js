// @format
import React from 'react';
import { Modal, View, ScrollView, TouchableOpacity, TouchableWithoutFeedback } from 'react-native';
import { PropTypes } from 'prop-types';

import { Carousel, Feedback } from 'app/components';
import { strings } from 'app/utils';
import { requestPermission, sendSentryError } from 'app/services';
import { NOTIFICATION_REJECTION_TOKEN } from 'app/services/firebase/notifications';

import { getOnBoardingData } from './ModalData';
import styles from './ModalProvider.styles';

export const ModalContext = React.createContext();

export default class ModalProvider extends React.Component {
  static propTypes = {
    acceptButtonText: PropTypes.string,
    children: PropTypes.node,
    error: PropTypes.object,
    isLoading: PropTypes.bool,
    isSubmitted: PropTypes.object,
    onTermsAcceptanceChange: PropTypes.func,
    postFeedback: PropTypes.func,
    termsAccepted: PropTypes.bool,
    text: PropTypes.string,
  };

  state = {
    content: '',
    data: [],
    isModalVisible: false,
    notificationPressed: false,
    notificationPressedText: '',
  };

  toggleVisibility = (isVisible) => {
    this.setState({ isModalVisible: isVisible });
  };

  requestPermission = async () => {
    let token;
    try {
      token = await requestPermission();
    } catch (err) {
      return sendSentryError('Error requesting permission in modal', err);
    }

    const notificationPressed = true;
    const notificationPressedText =
      token === NOTIFICATION_REJECTION_TOKEN
        ? 'You can enable notifications later from your Settings screen'
        : 'Notifications enabled!';

    // if this was successful, show notification success button
    this.setState({
      data: getOnBoardingData(this, notificationPressed, notificationPressedText),
      notificationPressed,
      notificationPressedText,
    });
  };

  openModal = (type) => {
    const { notificationPressed, notificationPressedText } = this.state;

    switch (type) {
      case strings.modalType.onboarding:
        this.setState({
          data: getOnBoardingData(this, notificationPressed, notificationPressedText),
        });
        break;
      case strings.modalType.feedback:
        this.setState({
          content: strings.modalType.feedback,
        });
        break;
    }
    this.toggleVisibility(true);
  };

  closeModal = () => this.toggleVisibility(false);

  renderContent = () => {
    if (this.state.content === strings.modalType.feedback) {
      const { error, isLoading, isSubmitted, postFeedback } = this.props;

      return (
        <Feedback
          error={error}
          isDismissButtonsVisible={true}
          isLoading={isLoading}
          isSubmitted={isSubmitted}
          onClose={this.closeModal}
          onSubmit={postFeedback}
        />
      );
    }
  };

  render() {
    const { children } = this.props;
    const {
      data,
      content,
      isModalVisible,
      notificationPressed,
      notificationPressedText,
    } = this.state;

    return (
      <ModalContext.Provider
        value={{
          notificationPressed,
          notificationPressedText,
          openModal: this.openModal,
          closeModal: this.closeModal,
          isModalVisible,
        }}
      >
        {children}
        {(!!data.length || !!content) && (
          <Modal
            style={styles.modal}
            animationType="slide"
            transparent
            visible={this.state.isModalVisible}
          >
            {data.length ? (
              <ScrollView style={styles.scrollView}>
                <View style={styles.contentContainer}>
                  <Carousel entries={data} fullWidth={false} />
                </View>
              </ScrollView>
            ) : (
              <TouchableOpacity
                style={styles.container}
                activeOpacity={1}
                onPressOut={this.closeModal}
              >
                <ScrollView style={styles.scrollView}>
                  <TouchableWithoutFeedback>
                    <View style={styles.contentContainer}>
                      <View>{this.renderContent()}</View>
                    </View>
                  </TouchableWithoutFeedback>
                </ScrollView>
              </TouchableOpacity>
            )}
          </Modal>
        )}
      </ModalContext.Provider>
    );
  }
}
