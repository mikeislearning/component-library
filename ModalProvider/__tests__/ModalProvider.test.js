/**
 * @format
 */

import React from 'react';
import { render } from 'react-native-testing-library';

import ModalProvider from '../ModalProvider';

describe('<ModalProvider />', () => {
  it('renders correctly', () => {
    const popup = render(<ModalProvider />);
    expect(popup).toMatchSnapshot();
  });
});
