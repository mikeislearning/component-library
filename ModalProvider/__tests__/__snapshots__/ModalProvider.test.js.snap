// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalProvider /> renders correctly 1`] = `null`;
